### ISSUE TITLE
Tor thread dies is unrecoverable after app goes to sleep/background

### ISSUE THREAD

**Issue by n8fr8** [OnionBrowser/iOS-OnionBrowser#103]
We need a way to recover gracefully when the Tor thread does kaput.

@chrisballinger any thoughts or tips on this?

**Comment by mtigas** [comment]
Current game plan:

This weekend I'll add something to my current fork of Tor.framework that sends `SIGNAL HALT` (see [spec](https://gitweb.torproject.org/torspec.git/tree/control-spec.txt#n406)). (Actually, Onion Browser 1.X does send a bunch of `SIGNAL HUP` to try to get Tor to reload config and change circuits and things. Should probably add that and the `NEWNYM` "use a new identity" signals too.)

That should give @tladesignz _something_ to work within a `applicationDidEnterBackground` callback (in terms of shutting down the existing Tor; then we can try to [start a new one](https://github.com/mtigas/OnionBrowser/blob/2.X/OnionBrowser/OBRootViewController.m#L206=L214)).

Still might not work, but I haven't experimented with this in forever (and may have done it wrong the last time I did) so it's worth giving it another shot with some fresh eyes on the problem.

**Comment by tladesignz** [comment]
So I thought a bit little bit about it:

The main problem is, as I understand it, that Tor is running in its own thread, the TORThread provided by the Tor.framework.

There's no mechanism in iOS for uncooperative threads to be killed from outside, since the thread could hang on to resources, mutexes etc. which then would never be properly handled. So, when you would actually be allowed that, and do that multiple times, the "mother"-process could at one point in time reach it's resource limits and crash.

So, what one needs to do inside a thread is, to check for the [cancelled](https://developer.apple.com/documentation/foundation/nsthread/1417366-cancelled?language=objc) flag and then close all resources and return gracefully in the [main](https://developer.apple.com/documentation/foundation/nsthread/1418421-main?language=objc) method.
Unfortunately, we can't do that, because we gave control to the C implementation of Tor [there](https://github.com/mtigas/Tor.framework/blob/master/Tor/TORThread.m#L66)

But if you get Tor to shut down as outlined, we should get control back. Actually, the thread should be done immediately after, because the `main` method would/should just end.

The nicest thing would be, if you manage to do that through overriding [cancel](https://developer.apple.com/documentation/foundation/nsthread/1411303-cancel?language=objc) in [TORThread](https://github.com/mtigas/Tor.framework/blob/master/Tor/TORThread.m).

It would be even better, if we could check, if Tor is still running, before we would call `cancel` on the thread, so we wouldn't need to restart it every time the app comes back to the front. Because it seemed to me that, often, Tor is still alive after I return to the application.

Is there a possibility to ask Tor about it's state and maybe timeout or something? Could get messy, though or even undoable, because we run into the same problem of beeing unable to kill a thread which is not responding and we just gave control to. Hmm.

In the end, it could also be, that Tor is not responding to any signal, because it's busy with waiting on some resource which iOS took away or something. In that case, Tor itself would need some patching for iOS to get it working.

**Comment by n8fr8** [comment]
This all sounds reasonable.

We could also try disabling Tor before background, and restarting when we return.

See this control port feature: https://www.torproject.org/docs/tor-manual.html.en

"DisableNetwork 0|1
When this option is set, we don’t listen for or accept any connections other than controller connections, and we close (and don’t reattempt) any outbound connections. Controllers sometimes use this option to avoid using the network until Tor is fully configured. (Default: 0)"

**Comment by tladesignz** [comment]
@n8fr8 Didn't you just mention some setting which makes Tor ignore time changes? As we discussed, it could also be, that Tor hangs because it is annoyed about time changes while it was suspended?

**Comment by tladesignz** [comment]
Ok, so I took the liberty and used some hours for this.
It mainly breaks, when I switch off the device, and on again. Could not reproduce it, when I just switched to another app.

Here's the error message I get, when the app was woken up again:

> 2017-08-31 13:40:14.744958+0200 OnionBrowser[12214:4958934] [] nw_socket_get_input_frames recvmsg(fd 34, 1024 bytes): [57] Socket is not connected

So there's some socket gone, after we return. As anticipated.

Trying to disable Tor like this, as @n8fr8 suggested didn't change anything at all:

```Swift
torController.setConfForKey("DisableNetwork", withValue: "1", completion: nil)
```

@mtigas, @n8fr8: Did I use that the right way?


I also tried to send `SIGNAL HALT` as suggested by @mtigas.
I exposed the internal method [`TORController#sendCommand`](https://github.com/mtigas/Tor.framework/blob/bce57fc7e48939ee9870125612724d0b5764ecf4/Tor/TORController.m#L437) and called it like this:

```Swift
        torController.sendCommand("SIGNAL HALT", arguments: nil, data: nil, observer: {
            (codes: Array<NSNumber>, lines: Array<Data>, stop: UnsafeMutablePointer<ObjCBool>) -> Bool in

            return true
        })
```

The result was, that the complete app process died immediately and without further notice.

@mtigas, @n8fr8: Did I use that the right way?

I mean, it seems so, since I actually got a reaction, but...

Do you guys have any other ideas how to pull this off?

**Comment by tladesignz** [comment]
So, I compared Onion Browser 1 with 2, which can't repair a broken Tor connection but at least handles the broken connection gracefully.

These are my findings:

- OB1 handles the connection lose in a UIWebView callback:
```Objective-C
- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
```

[WebViewController.m#L581](/mtigas/OnionBrowser/blob/1.X/OnionBrowser/OnionBrowser/WebViewController.m#L581)
[WebViewController.m#L612](/mtigas/OnionBrowser/blob/1.X/OnionBrowser/OnionBrowser/WebViewController.m#L612)

The equivalent in OB2's Endless would be here: [WebViewTab.m#L464](/mtigas/OnionBrowser/blob/2.X/Endless/WebViewTab.m#L464)

Unfortunately, in OB2, we never come that far on app resume.

The furthest, I understand, we come is [UIApplicationDelegate#applicationWillEnterForeground](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623076-applicationwillenterforeground), which is a callback, we currently don't use and I just added for debugging purposes.

Everything else I see is no log and a lot of obfuscated debug garbage. On retries, it even crashes on different threads, some which seem to do nothing else than waiting and then getting released just before the crash. 

What did catch my eye, though, was the cache cleanup of Endless, which is quite aggressive and also deletes Tor related temporary files, which can be disabled by commenting [CookieJar.m#L170](/mtigas/OnionBrowser/blob/2.X/Endless/CookieJar.m#L170).
However, this didn't make a difference regarding the crashes.

I'm totally at a loss, here. Logging could maybe help. @mtigas can we do [this pull request](/mtigas/Tor.framework/pull/1)? Do you see any problems?

**Comment by tladesignz** [comment]
Finally found the issue a while ago. Is fixed and 2.0 is released, so I close this.
